src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -2020,253 +2175,86 @@ public void observe(@NonNull SCMHead head, @NonNull SCMRevision revision) throws
-                    return;
-                }
-                try (ChildNameGenerator.Trace trace = ChildNameGenerator.beforeCreateItem(
-                        MultiBranchProject.this, encodedName, branch.getName()
-                )) {
-                    if (getItem(encodedName) != null) {
-                        throw new IllegalStateException(
-                                "JENKINS-42511: attempted to redundantly create " + encodedName + " in "
-                                        + MultiBranchProject.this);
-                    }
-                    project = _factory.newInstance(branch);
-                }
-                if (!project.getName().equals(encodedName)) {
-                    throw new IllegalStateException(
-                            "Name of created project " + project + " did not match expected " + encodedName);
-                }
-                // HACK ALERT
-                // ==========
-                // We don't want to trigger a save, so we will do some trickery to ensure that observer.created(project)
-                // performs the save
-                BulkChange bc = new BulkChange(project);
+        private SCMRevision getLastSeenRevision(@NonNull P project, SCMRevision scmLastBuiltRevision) {
+            SCMRevision scmLastSeenRevision = _factory.getLastSeenRevision(project);
+            if (scmLastSeenRevision == null && scmLastBuiltRevision != null) {
+                scmLastSeenRevision = scmLastBuiltRevision;
                 try {
-                    if (project.getDisplayNameOrNull() == null && !rawName.equals(encodedName)) {
-                        project.setDisplayName(rawName);
-                    }
-                } catch (IOException e) {
-                    // ignore even if it does happen we didn't want a save
-                } finally {
-                    bc.abort();
-                }
-                // decorate contract is to ensure it dowes not trigger a save
-                _factory.decorate(project);
-                // ok it is now up to the observer to ensure it does the actual save.
-                observer.created(project);
-                if (isAutomaticBuild(source, head, revision, null, null, listener)) {
-                    scheduleBuild(
-                            _factory,
-                            project,
-                            revision,
-                            listener,
-                            rawName,
-                            causeFactory.create(source),
-                            revisionActions
-                    );
-                } else {
-                    listener.getLogger().format("No automatic builds for %s%n", rawName);
-                }
-                try {
-                    _factory.setLastSeenRevisionHash(project, revision);
+                    _factory.setLastSeenRevisionHash(project, scmLastBuiltRevision);
```

Comment:
A _getter_ that mutates data isn't very intuitive at first glance. It's a private method, so not a big deal. But maybe just call it `lastSeenRevision(...)` for a tiny bit ore clarity?